### PR TITLE
Terminal+TerminalSettings: Overhaul Terminal Settings and extract it into its own application

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
Functionality is practically unchanged. Now, TerminalSettings is its own, massively improved, standalone application, just as the other settings applications.

image

image

### CODE REVIEW

File: Userland/Applications/Terminal/main.cpp [hunk 1]
```diff
@@ -4,9 +4,10 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include "LibConfig/Listener.h"
+#include "LibGfx/FontDatabase.h"
```
Comment (linusg) [inline Userland/Applications/Terminal/main.cpp:8]: Please use `<>`-style includes

File: Userland/Applications/Terminal/main.cpp [hunk 2]
```diff
@@ -45,6 +42,53 @@
...
+            }
+        } else if (group == "Text" && key == "Font") {
+            auto font = Gfx::FontDatabase::the().get_by_name(value);
+            if (font.is_null())
+                font = Gfx::FontDatabase::default_fixed_width_font();
+            m_parent_terminal.set_font_and_resize_to_fit(*font);
+            m_parent_terminal.window()->resize(m_parent_terminal.size());
+        }
+    }
+    virtual void config_i32_did_change(String const& domain, String const& group, String const& key, i32 value)
```
Comment (linusg) [inline Userland/Applications/Terminal/main.cpp:77]: Please separate these with a newline

File: Userland/Applications/TerminalSettings/TerminalSettingsWidget.cpp [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+ * Copyright (c) 2018-2021, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "TerminalSettingsWidget.h"
+#include "AK/Assertions.h"
+#include "LibGUI/Button.h"
+#include "LibGfx/Font.h"
+#include "LibGfx/FontDatabase.h"
```
Comment (linusg) [inline Userland/Applications/TerminalSettings/TerminalSettingsWidget.cpp:11]: Same as before

File: Userland/Applications/TerminalSettings/TerminalSettingsWidget.cpp [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+void TerminalSettingsMainWidget::write_back_settings() const
+{
+    dbgln("{}, {}", m_original_max_history_size, stringify_bell(m_original_bell_mode));
+    Config::write_i32("Terminal", "Terminal", "MaxHistorySize", static_cast<i32>(m_original_max_history_size));
+    Config::write_string("Terminal", "Window", "Bell", stringify_bell(m_original_bell_mode));
+}
+
+void TerminalSettingsMainWidget::cancel_settings()
+{
+    dbgln("Cancel called!");
```
Comment (linusg) [inline Userland/Applications/TerminalSettings/TerminalSettingsWidget.cpp:171]: leftover from debugging?

File: Userland/Applications/TerminalSettings/TerminalSettingsWidget.cpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+void TerminalSettingsMainWidget::apply_settings()
+{
+    m_original_max_history_size = m_max_history_size;
+    m_original_bell_mode = m_bell_mode;
+    write_back_settings();
+}
+void TerminalSettingsMainWidget::write_back_settings() const
+{
+    dbgln("{}, {}", m_original_max_history_size, stringify_bell(m_original_bell_mode));
```
Comment (linusg) [inline Userland/Applications/TerminalSettings/TerminalSettingsWidget.cpp:164]: leftover from debugging?

**Review by linusg — COMMENTED** [review]
very nice :^)

**Comment by awesomekling** [comment]
Those explanation labels seem unnecessary, almost condescending.

File: Userland/Applications/TerminalSettings/TerminalSettingsView.gml [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        title: "Terminal Font"
+        fixed_height: 100
+
+        layout: @GUI::VerticalBoxLayout {
+            margins: [16, 8, 8]
+            spacing: 16
+        }
+
+        @GUI::Label {
+            text: "By default, this is the system-wide fixed-width font."
```
Comment (awesomekling) [inline Userland/Applications/TerminalSettings/TerminalSettingsView.gml:41]: This is very opaque. How do I know if it's in "default font" state, or if I've specifically assigned a font that happens to be the default?
Instead, we could have a "Use system default" checkbox, that disables the font selector if checked.

File: Userland/Applications/TerminalSettings/TerminalSettingsView.gml [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        title: "Color Scheme"
+        fixed_height: 100
+
+        layout: @GUI::VerticalBoxLayout {
+            margins: [16, 8, 8]
+            spacing: 16
+        }
+
+        @GUI::Label {
+            text: "Depending on your theme, color schemes may not be available."
```
Comment (awesomekling) [inline Userland/Applications/TerminalSettings/TerminalSettingsView.gml:83]: If color schemes are not available due to `$reason`, we can show a label explaining why.
If they are available, we don't need to show anything.

File: Userland/Applications/TerminalSettings/TerminalSettingsView.gml (hunk 6 above)
Comment (kleinesfilmroellchen) [reply Userland/Applications/TerminalSettings/TerminalSettingsView.gml:41]: That's a great idea!

**Comment by kleinesfilmroellchen** [comment]
@awesomekling I tried to do a similar level of explanations as in MailSettings, I think the argument goes for both. But of course if you feel strongly, I'll remove the labels.

**Comment by awesomekling** [comment]
> @awesomekling I tried to do a similar level of explanations as in MailSettings, I think the argument goes for both. But of course if you feel strongly, I'll remove the labels.

Hmm, the MailSettings one are a bit over-the-top as well. I don't feel super-strongly (but maybe medium-strongly).
In general, the system UI should assume the user to be competent. :)

**Comment by kleinesfilmroellchen** [comment]
In that case, I'll dial it back.

**Pull request merged**